delta_go/go_base.py

```diff
@@ -0,0 +1,253 @@
+
+    # ...and retaking ko is always illegal
+    if ko is not None:
+        legal_moves[ko] = False
+    # Concat with pass move
+    return np.arange(BOARD_SIZE**2 + 1)[legal_moves.ravel().tolist() + [True]]
+
+
+def play_move(state, move: int, color=None, mutate=False):
+    # Obeys CGOS Rules of Play. In short:
+    # No suicides
+    # Chinese/area scoring
+    # Positional superko (this is very crudely approximate at the moment.)
+    if color is None:
+        color = state.to_play
+
+    pos = state if mutate else copy.deepcopy(state)
+
+    coord = int_to_coord(move)
+    if coord is None:
+        pos = pass_move(state, mutate=mutate)
+        return pos
+
+    if not is_move_legal(coord, state.board, state.ko):
+        raise IllegalMove(
+            f'{"Black" if state.to_play == BLACK else "White"} coord at {coords.to_gtp(coord)} is illegal: \n{state}'
+        )
+
+    potential_ko = is_koish(state.board, coord)
+
+    place_stones(pos.board, color, [coord])
+    captured_stones = pos.lib_tracker.add_stone(color, coord)
+    place_stones(pos.board, EMPTY, captured_stones)
+
+    opp_color = color * -1
+
+    new_board_delta = np.zeros([BOARD_SIZE, BOARD_SIZE], dtype=np.int8)
+    new_board_delta[coord] = color
+    place_stones(new_board_delta, color, captured_stones)
+
+    if len(captured_stones) == 1 and potential_ko == opp_color:
+        new_ko = list(captured_stones)[0]
+    else:
+        new_ko = None
+
+    pos.ko = new_ko
+    pos.recent_moves += (PlayerMove(color, move),)
+
+    # keep a rolling history of last 7 deltas - that's all we'll need to
+    # extract the last 8 board states.
+    pos.board_deltas = [new_board_delta.reshape(1, BOARD_SIZE, BOARD_SIZE)] + pos.board_deltas[:6]
+    pos.to_play *= -1
+    return pos
+
+
+def game_over(recent: Tuple[PlayerMove, ...]) -> bool:
+    return len(recent) >= 2 and recent[-1].move == PASS_MOVE and recent[-2].move == PASS_MOVE
+
+
+def score(board: np.ndarray, komi: float, return_both_colors: bool = False) -> Union[float, Tuple]:
+    """Return score from Black's perspective.
+
+    If White is winning, score is negative.
+    return_both_colors: set to true to return a tuple of (black_score, white_score)
+                        rather than white score = -black_score
+    """
+    working_board = np.copy(board)
+    while EMPTY in working_board:
+        unassigned_spaces = np.where(working_board == EMPTY)
+        c = unassigned_spaces[0][0], unassigned_spaces[1][0]
+        territory, borders = find_reached(working_board, c)
+        border_colors = {working_board[b] for b in borders}
+        X_border = BLACK in border_colors
+        O_border = WHITE in border_colors
+        if X_border and not O_border:
+            territory_color = BLACK
+        elif O_border and not X_border:
+            territory_color = WHITE
+        else:
+            territory_color = UNKNOWN  # dame, or seki
+        place_stones(working_board, territory_color, territory)
```

Comment:
I would love it if this bit I've highlighted could be a separate function - maybe call it `get_territory_board` or something? 

Why? KataGo trained an extra neural network head to predict the final board location colours, this allows us to calculate that so we can use it as a target